design-proposals/klip-8-interactive-queries.md

```diff
@@ -131,13 +131,130 @@ Additionally:
 * Key movement should be handled transparently: if a query launches and a targeted key physically moves, KSQL should redirect the query to the appropriate KSQL server to obtain the correct result.
 * For interactive queries against windowed queries, the user may specify a time range via the `WINDOWSTART` and `WINDOWEND` psuedocolumns. For example, `SELECT * FROM windowed_aggregate WHERE ROWKEY = 'rowkey' AND WINDOWSTART >= start_ts AND WINDOWEND < end_ts`
 
-## Open questions
+---
+
+# Syntax
+
+All KSQL query syntax currently yields streaming results because KSQL only supports streaming queries. As a result, adding support for point-in-time queries necessitates a means to unambiguously differentiate between streaming and finite query results. The most straightforward approach to this is to use syntax to indicate streaming versus PIT queries. And since there are only two fundamental query forms, it is only necessary to add KSQL syntax to represent one of them.
+
+We therefore propose making minimal syntax changes to represent *streaming* queries. The principal reason for using new syntax for streaming queries is that PIT queries can then use syntax that is as close to the SQL standard as possible, making KSQL interoperable with tooling that integrates with SQL-based systems (i.e. dashboarding frontends).
+
+After introducing any syntax changes, KSQL's syntax must make sense in each of the following contexts:
+
+* Streaming query on a stream
+* Streaming query on a table
+* Point-in-time query on a stream
+* Point-in-time query on a table
+* Transient queries defined by any of the above
+* CTAS/CSAS defined by any of the above
+
+*Note that not all of these query forms need to be supported (initially, or ever), but a sensible design should allow for all of them with minimal ambiguity.*
+
+The proposed syntax changes will now be specifically described.
+
+## Proposed syntax changes
+
+Streaming queries may be identified via the `EMIT CHANGES` query modifier:
+
+```sql
+SELECT ... FROM stream EMIT CHANGES;
```

Comment:
Since its not clear if we will support PIT on streams, can we just add the EMIT CHANGES to tables alone for now?  This will avoid changes to the existing streaming queries on streams (which I guess is the vast majority of queries).

Reply:
> To confirm, without EMIT CHANGES , this query would become point in time? i.e will return all events from the stream.

This wouldn't necessarily be the case and would depend on a couple of things:

1. Do we want to support PIT queries on streams?
2. What are the semantics of a PIT query on a stream? Where do we begin consuming from? Where do we end?

> any way around this, migration plans? I don't think this is called out in the section below?

I don't think there is a way around this. If we're adding syntax to indicate streaming queries, that will effectively invalidate all existing KSQL queries. It seems that the only alternative would be to identify PIT queries with new syntax so that all existing KSQL queries continue to work as expected, but then we lose the benefit of keeping PIT queries standard.

The best we'll likely be able to do here is outline a brute-force migration approach: create all queries in a new KSQL and consume as much topic history as you need/want. Perhaps it would be worth including a simple query rewrite tool that can take a `.sql` file of "legacy" streaming queries and rewrite them to use `EMIT CHANGES` where appropriate. Honestly we should probably have some kind of [dump](https://www.postgresql.org/docs/12/app-pgdump.html) tool anyways, so maybe this could justify creating one. `dump` would essentially dump all KSQL objects as DDL statements:

```
$ ksql-dump > dump.sql
$ cat dump.sql
CREATE STREAM s0 (x integer) WITH (kafka_topic='kt', value_format='delimited');
CREATE TABLE t0 AS SELECT x, count(*) FROM s0 GROUP BY x;
```

It could detect a legacy KSQL version and add `EMIT CHANGES` where necessary.

You could then reload your environment into a new KSQL using something like (`-f` flag is fake!):
```
$ ksql -f dump.sql
```

Anyways, I will add something to this KLIP calling all of this out. Thanks for bringing it up @vinothchandar!

Reply:
A previous suggestion (not fully captured in the rejected alternatives below) was to be context dependent. 

In particular, have all queries on today's streams remain as they are. By default, `select * from table` is a point in time query. `CTAS` statements on tables are streaming by default. If you want `select * from table` to be streaming, you would have to add `EMIT CHANGES`. 

The benefits here is that we don't break existing users, we keep the syntax succinct for the most common use cases, we address the biggest surprise of current users that `select * from table` returns an unbound stream. 

It also reserves the right to introduce the new syntax for different categories over time, incrementally. 

> To confirm, without EMIT CHANGES , this query would become point in time? 

What would be the failure mode in v1 where we don't support PIT on streams. Just fail the query with an error ? We should write up the UX and failure modes when people try things not supported. 

Also, from a cost perspective, every single demo video, tutorial, and quickstart will have to be re-written with the current proposal. Worth noting the `Cons` somewhere too, so people can weigh the proposal accordingly.

Reply:
> A previous suggestion (not fully captured in the rejected alternatives below) was to be context dependent.

@apurvam thanks for the reminder, I will add this to the *Considered approaches* section. I've added a **TODO** checklist to the description of this PR for action items arising from the discussion here.

> In particular, have all queries on today's streams remain as they are. By default, select * from table is a point in time query. CTAS statements on tables are streaming by default. If you want select * from table to be streaming, you would have to add EMIT CHANGES.

I fear that context-dependent semantics could very quickly become confusing and/or burdensome for users. The understanding that any query will always have the same behavior regardless of context would be a big usability win in my opinion. This is especially important given that users will often be prototyping with transient queries in the CLI until they get what they want for declaration via CTAS/CTAS.

That being said, it is hard to deny the pragmatism of what you propose, as it has a pretty big upside (backward compatibility). Another way to think about this is that the following `SELECT` queries all return equivalent streams:

```sql
CREATE STREAM s0 AS SELECT * FROM s1;
SELECT * FROM s1 EMIT CHANGES;
CREATE STREAM s0 AS SELECT * FROM s1 EMIT CHANGES; -- EMIT CHANGES is a noop here
```

In a "pure" implementation (not context dependent) the following would create a stream initially populated with the one-time result of the corresponding PIT query on a stream `s1`:

```sql
CREATE STREAM s0 AS SELECT * FROM s1;
```

And this (initial population of a stream) is something that seems like an exceptionally rare requirement anyways. I do still prefer the explicit approach (requiring `EMIT CHANGES`) for the reasons I noted above, as well as one additional reason: KSQL is strongly typed, and that warrants expectations around looking at an expression and knowing what it will return.  

> What would be the failure mode in v1 where we don't support PIT on streams. Just fail the query with an error ? We should write up the UX and failure modes when people try things not supported.

I think a simple, descriptive error would be fine here. For example,

```
ksql> SELECT * FROM stream;
Point-in-time queries on streams are currently not supported
```

> Also, from a cost perspective, every single demo video, tutorial, and quickstart will have to be re-written with the current proposal.

This is a fair point, although adding `EMIT CHANGES` where applicable does seem like a manageable undertaking. We will likely be doing a lot of docs work with the addition of PIT queries anyways. Ultimately I feel that establishing the right interface is the priority here, regardless of upfront docs/maintenance cost (within reason!).

> Worth noting the Cons somewhere too, so people can weigh the proposal accordingly.

Good call, I've also added this to the **TODOs**.

Reply:
Thanks @derekjn. I think that until we clearly want to support PIT queries on streams, and thus have a need to break backward compatibility, we should not change the current semantics.

Also, the mental model becomes simpler. You can _only_ do continuous queries on streams. Nothing changes from today.

With regard to tables, for CTAS queries, the default is to feed the changelog of the source into the destination table with the modifications specified in the query. All other queries on tables default to point in time queries. 

Yes this is context dependent, and yes it could present a usability problem, but the question is whether the underlying concepts are simple enough that people can reason about them with a bit of training. It might be worth taking an incremental step and finding out whether it works or not before making a big leap. 

cc @blueedgenick , @rmoff . It would be good to get your input on this:  you have the most exposure to KSQL users in the wild. Also, a lot of your material is at risk of needing significant rework with this proposal.

Reply:
Only had time to skim, but...

I think making it context dependent should be avoided.

1. I think it'll be confusing to users, who expect a persistent query to work the same as a transient one.
2. It means the syntax can no longer support creating a static table in the future.
i.e. `CREATE TABLE AS SELECT * FROM A_TABLE` could be used in the future to create a static snapshot table, should we want such a thing.

Reply:
Not to heap on, but @big-andy-coates's point around (1) is really important. Transient queries are the means for experimenting before you roll a statement up into a CSAS/CTAS. Diverging behavior there would really detract from that mode of working.

Agree that taking a hit on changing previous docs/materials is unfortunate, but the flip side is that if we think we have a better model that happens to be backward-incompatible, making the change now is a one-time pain vs. having to carry a suboptimal model indefinitely.

Reply:
> I think it'll be confusing to users, who expect a persistent query to work the same as a transient one.

A very quick (10 minute) search showed at least 4 different people being confused by the current behavior for `select * from table` to be continuos, even they are comfortable with CTAS. 

https://github.com/confluentinc/ksql/issues/1149
https://github.com/confluentinc/ksql/issues/1876
https://github.com/confluentinc/ksql/issues/2477
https://github.com/confluentinc/ksql/issues/1336

I really feel we need an opinion from the field facing folks (who actually interact with real users) before me make an assumption about what is intuitive and what is not. 

I also don't think we should be breaking all queries just for anticipated future functionality (snapshot CTAS queries, point queries on streams). Seems like a poor trade off to me. 

If we want to make incremental changes we should start doing that now. If we have data to show that the new behavior is actually what users want, we should use that. Otherwise we should make the minimal set of changes with the data we actually have from a product which has been out there for 2 years.

Reply:
> A very quick (10 minute) search showed at least 4 different people being confused by the current behavior for select * from table to be continuos, even they are comfortable with CTAS.

I understand the implication here that users assume that CTAS is continuous but expect `SELECT * FROM table` to be PIT. But I also feel that the proposed syntax does specifically address this confusion. With the proposal, `SELECT * FROM table` would be PIT and a streaming CTAS would require an explicit `EMIT CHANGES` modifier so any ambiguity would essentially be eliminated in my opinion.

> Otherwise we should make the minimal set of changes with the data we actually have from a product which has been out there for 2 years.

I think it's worth noting that this proposal involves adding a fundamentally new capability to KSQL, so I'm not sure that history helps us here either way. I also believe that choosing to solve this problem with minimal incremental changes is just as significant of a decision as choosing the `EMIT CHANGES` approach. They're both conscious decisions with long-term implications, and personally I don't feel that the significance of whatever approach we take here is necessarily related to how similar the approach is to KSQL today.

Reply:
> I think it's worth noting that this proposal involves adding a fundamentally new capability to KSQL, so I'm not sure that history helps us here either way. 

I think we can agree to disagree. Data transformation will remain a core part of KSQL even with the additions we are making for point in time queries. What we have learnt about this should have bearing on how we design for the future. More importantly, we have an existing user base, many of whom have asked for precisely the feature being discussed here (see the number of reactions on #530)). So having a solution which is palatable for them (ie. doesn't break everything) in the first iteration seems like a reasonable objective.

I still can't see the net positive on changing the semantics for `select * from some_stream` to become a PIT query on streams. It will be unsupported for the forseeable future and maybe forever.

I can buy the argument to now require `EMIT CHANGES` for `CTAS` statements to keep  today's CTAS behavior while enabling point in time queries on tables through regular selects. I agree that omitting this change _could_ confuse users. I would still prefer to avoid it because it will be redundant syntax for the foreseeable future, and it breaks existing statements. We simply don't know that we want 'snapshot' CTAS queries. If such use cases present themselves, we can introduce the requirement for this clause at that point. 

Thus my vote would be be to _only_ change the semantics of `select * from table`  to become point in time right now: this enables the functionality we want to add, and it matches users' expectations for that statement based on the data we have (samples of which are referenced above, but field feedback is especially important here). To get continuous queries on tables, users will have to do `select * from table EMIT CHANGES`. 

Everything else should remain the same until we have good reasons to change it. For instance, if everybody starts getting confused about why a `select * from some_table` doesn't behave the same as `create table as select * from some_table`, that may be a good reason to require `EMIT CHANGES` for the `CTAS` variant.

Reply:
>>I also don't think we should be breaking all queries just for anticipated future functionality (snapshot CTAS queries, point queries on streams). Seems like a poor trade off to me.

+1 I also feel strongly about not supporting EMIT CHANGES syntax on streams and leave the current behavior. Alternatively, we can fully flesh out the long term roadmap/design here (PIT on Streams, should table be the only entity, and all other open discussion points) and ideally batch up other breaking syntax changes also along with this.   @apurvam pointed out, the overhead of re-training, re-doing docs and migrating queries is simply too high to repeat again.